### ISSUE TITLE
Identifier - a suggested enhancement to its definition

### ISSUE THREAD

**Issue by damianmoz** [chapel-lang/chapel#16310]
There is:
```
    An identifier in Chapel is a sequence of characters that starts with a lowercase
    or uppercase letter or an underscore and is optionally followed by a sequence of
    lowercase or uppercase letters, digits, underscores, and dollar-signs.
```
And this works effectively.

There was:
```
    The items of information which a ?????? program handles are identified by names.
    ...
    A large name is an upper-case letter, possibly followed by a string of letters and
    digits, possibly followed by one or more primes. A, Ab3C', ABC, A'', are large names.
```
And that worked effectively and robustly.

There maybe:
```
    An identifier in Chapel is a sequence of characters that start with a lowercase
    or uppercase letter or an underscore and is optionally followed by a sequence
    of lowercase or uppercase letters, digits, underscores, and dollar-signs, possibly
    followed by one or more primes.
```
Would that work or break anything?  It would make expressing a lot of mathematical algorithms in Chapel just SO, SO MUCH EASIER!

Does Chapel use a prime or single quote for anything other that strings. Chapel is not Ada so it does not use T'first and T'last to extract items from a range.

Does this rock the boat too much? It is not like it is something new or unproven.

**Comment by mppf** [comment]
@damianmoz - it took me a while to figure out but I think you are saying you would like any number of single quote characters (`'`) to be something you can append to a name. (I kept thinking *prime* meant something else, for some reason).

I can't think of a situation in the parser where a string is allowed immediately after an identifier, as with `f'hello'` - with an intervening space or not. As a result I don't think this would change the behavior of any program that compiles. (In other words, I think it would be a change that preserves backwards compatibility).

So, it would be fine with me!

**Comment by lydia-duncan** [comment]
> I can't think of a situation in the parser where a string is allowed immediately after an identifier, as with `f'hello'` - with an intervening space or not.

Isn't that one of the ways we support bytes, e.g. `b'foo'`?

**Comment by mppf** [comment]
> Isn't that one of the ways we support bytes, e.g. `b'foo'`?

Good point. However I will not be worried if one can't name a variable specifically `b'`.

**Comment by bradcray** [comment]
> but perhaps that makes the proposal less appealing.

It seems like it would make the proposal slightly less appealing, and would also prevent us from adding new string literal formats going forward (not that I'm aware we have plans to do so, but I didn't anticipate us adding `b'...'` a few years ago either).

(I'd personally be happy to see the `c'...'` string literal format go away as we move away from c_strings).

**Comment by damianmoz** [comment]
As Chapel uses the more Algol60-like concept of a string and not the BCPL-like concept of a 1D array of **char**, I would think there is no reason to allow a single quote to encapsulate strings or bytes within the language. If you look at the current Chapel documentation for bytes, it only uses double quotes, i.e. ".  The same for strings. I assume that this is deliberate. I have personally started to try and avoid using a single quote anywhere in my own programs and am trying to enforce this with others. Some people even think that **:** are quotation marks and call ASCII 0x2c, the so called single quote **'**, by the name _prime_.

But it is way above my station to push that barrow.

Anyway, I could always ask about the availability of ASCII 0x60, the so-called back-quote **`** for my purpose if its more popular sibling, the single quote, is off-limits. I have not seen how that works in practice and unlike my proposal, there is no proven precedent.

This is not a big deal.

**Comment by mppf** [comment]
@damianmoz - would it bother you if `b'`, `b''`, and `b'''` could not be used as a variable name but all other names could use `'` e.g. `_b'` or `bb'` or `x'''` ?

**Comment by damianmoz** [comment]
It would. But only from the perspective that it does not seem clean. And that means it will stab either myself or somebody else in the back at a later stage. Whatever we do must have no exceptions. Otherwise we are looking for trouble.

**Comment by mppf** [comment]
I suppose we could instruct the lexer to treat `b''` as a variable name but `b""` and `b' hi '` as bytes strings. That just moves the exception from one about variable names to making `b''` not be the empty bytes string (but one could use `b""` for that).

**Comment by bradcray** [comment]
I think the path @mppf is on in the previous comment is fraught with difficulties (having tried to think down the same path myself last night).  Is `b' + c'` a variable named `b'` plus a variable named `c'` or is it the byte string " + c"?

> I would think there is no reason to allow a single quote to encapsulate strings or bytes within the language. If you look at the current Chapel documentation for bytes, it only uses double quotes, i.e. ". The same for strings. I assume that this is deliberate.

Grepping the code base, the single quote form is used a lot, unfortunately for this proposal.  Supporting both forms has the benefits of simplifying the ability to create strings containing `"` or `'` (but not both) without having to use escapes.  And it has the probably more significant benefit of appealing to Python / scripting language programmers.  And we're trying to avoid making additional breaking language changes at this point (where extending identifiers is not a breaking change, but disabling single-quote strings or byte strings is).

Like Damian, I'd similarly almost counter-proposed using a different single quote form, but was worried that it would either render in an ugly way (e.g., A`) or be sufficiently difficult to type as to be unattractive (e.g., A’ where that's meant to be a unicode smart quote).  But if it's attractive to users, I'm open to those with only slight trepidation.

**Comment by damianmoz** [comment]
I am seeing text rendered with a greyscale background box. That helps readability. How do you achieve this as it is neither bold or italic.

**Comment by mppf** [comment]
@damianmoz - in the github flavored markdown here - you can use the backtick for that e.g.
```
`hi`
```

renders as `hi`

**Comment by damianmoz** [comment]
Thanks. So it seems that what I call a back quote, you call a back tick. 

I think that the back tick **A`** would work for me. Let's stay ASCII please. No Unicode.
```chapel
y`` = y` * x
```
or even better if we incorporate an earlier proposal of mine
```chapel
y`` := y` * x
```
That looks good for me. I am happier with that than I thought I would be. But I use a US keyboard with Australian English display. I wonder how it looks on a non-English LOCALE. Is there some Chapel user with a Chinese or Japanese LOCALE out there who could verify how this renders?

**Comment by damianmoz** [comment]
Sorry, I hit the wrong button. How do I reopen this?

**Comment by lydia-duncan** [comment]
I think we're the only ones that can :)

**Comment by damianmoz** [comment]
So the question is how well the ASCII back-tick or back-quote sits with others as a concept and whether it renders sufficiently clearly on the screen of the programmer and within any documentation to be obviously differently to what we all call a single quote. I notice on my
own SSH window into my Linux machine, the back-tick is rendered with what appears to be one less pixel (so it looks less dense) that the single quote. But it is still quite readable and looks clearer than adding say an underscore which is what I currently often use if I want to refer to the numerically scaled variant of a existing variable. 

We await feedback I guess.

**Comment by bradcray** [comment]
I'm obviously not the target audience, but I find back-tick sufficiently visually different from single-quote to not be overly confusing (esp. since I expect them to come up in fairly different contexts in general).  So if it's sufficiently attractive to users like Damian, I'm open to using it in this way.

One thing about it I puzzle over a bit is whether the back-ticks should only be permitted at the end of an identifier, which I think would be the first case of such a "terminating only" character.  We could alternatively permit them to be anywhere other than the starting character, but then have the style guide dissuade people from putting them elsewhere.  Or we could introduce a class of characters that could only appear at the end of an identifier.  I take it this is what Fortran did?  (I'm assuming that was the origin of Damian's original quote?)

**Comment by damianmoz** [comment]
Fortran is not that intelligent!!

I deliberately followed exactly what CPL did. This means it is being only allowed as trailing what we all think of as identifier names. That way, I had prior usage upon which to fall back to claim usefulness and productivity from an end user's perspective, the ease (or otherwise) of incorporation into the language from a compiler writer's perspective.  And I know people still around who had both used it and implemented it and to varying levels of involvement in. So there is first-hand experience on which to draw. There seemed to be no downside to this feature unlike some other language features seen in the past for a raft of languages. And neither of these were numerical people so they were slightly more neutral than myself.
